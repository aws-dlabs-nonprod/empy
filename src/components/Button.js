import React, { Component } from 'react';
import styled from 'styled-components';
import ReactSVG from 'react-svg';
import Media from '@style/media';

import { StyledLoader } from '@components/Loader';
import Tip, { StyledTip } from '@components/ButtonTip';

class Button extends Component {

    static defaultProps = {
        tip: null,
        toggleTip: null,
        tipPosition: 'left',
        isToggled: false,
        secondary: false,
        disabled: false,
        onClick: () => {}
    };

    constructor(props) {
        super(props);
    }

    handleClick = (e) => {
        e.preventDefault();

        this.props.onClick();
    }

    render() {
        const {
            children,
            className,
            icon,
            toggleIcon,
            tip,
            toggleTip,
            isToggled,
            disabled,
            tipPosition,
            tipBelow,
            theme
        } = this.props;

        const tipContent = isToggled ? (toggleTip || tip) : tip;
        const iconOption = isToggled ? (toggleIcon || icon) : icon;

        return (
            <button disabled={ disabled } aria-label={ tipContent } className={ className } onClick={ this.handleClick }>
                { iconOption && <ReactSVG src={ iconOption } /> }

                { children }

                { tipContent &&
                <Tip theme={ theme } tipPosition={ tipPosition } tipBelow={ tipBelow }>
                    <p>{ tipContent }</p>
                </Tip>
                }
            </button>
        );
    }
}

const BaseButton = styled(Button)`
    background: ${props => props.theme.colourPrimary};
    border-radius: ${props => props.theme.radius};
    box-sizing: border-box;
    color: ${props => props.theme.colourForeground};
    cursor: pointer;
    display: flex;
    justify-content: center;
    font-size: 1.6rem;
    height: 4.8rem;
    border: ${props => `0.1rem solid ${props.theme.colourPrimary}`};
    margin: 0 0 2rem 0;
    outline: none;
    padding: 0 1.8rem;
    position: relative;
    width: 100%;
    z-index: 3;
    transition: background 0.2s ease, color 0.2s ease;

    ${Media.tablet`
        font-size: 1.8rem;
        padding: 0 2rem;
        width: 100%;

        &:hover {
            background: ${props => props.theme.colourPrimaryHover};
            
            svg {
                fill: ${props => props.theme.colourForeground};
            }

            ${StyledTip} {
                opacity: ${props => props.tip ? 1 : 0};
                transform: translateY(0);
            }
        }
    `}
`;

const SecondaryButton = styled(BaseButton)`
    background: ${props => props.theme.colourForeground};
    color: ${props => props.theme.textPrimary};

    ${Media.tablet`
        &:hover {
            color: ${props => props.theme.colourForeground};

            ${StyledLoader} {
                circle {
                    stroke: ${props => props.theme.colourForeground};
                }
            }
        }
    `}

    ${StyledLoader} {
        circle {
            stroke: ${props => props.theme.colourPrimary};
        }
    }
`;

const IconButton = styled(BaseButton)`
    background: ${props => props.theme.colourForegroundHover};
    border-color: ${props => props.isToggled ? props.theme.colourPrimary : props.theme.colourDivider};
    padding: 1.4rem 1.3rem;

    ${Media.tablet`
        padding: 1.4rem 1.3rem;

        &:hover {
            background: ${props => props.theme.colourForeground};

            svg {
                fill: ${props => props.theme.colourPrimary};
            }
        }
    `};

    /* Icons */
    svg {
        fill: ${props => props.theme.colourPrimary};
        height: 2rem;
        position: relative;
        width: 2rem;
    }
`;

const SmallIconButton = styled(IconButton)`
    height: 3rem;
    width: 3rem;

    ${Media.tablet`
        padding: 0.4rem;
        height: 3rem;
        width: 3rem;
    `};
`;

export default BaseButton;

export {
    BaseButton as Button,
    SecondaryButton,
    IconButton,
    SmallIconButton
};
