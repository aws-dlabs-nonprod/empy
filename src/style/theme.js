export default {
    textPrimary: '#2D373E',
    textSecondary: '#E7E6E6',
    textBody: '#1A1A1A',
    textDark: '#621A4B',

    colourPrimary: '#D5002B',
    colourPrimaryHover: '#e24d6b',
    colourPrimaryTransparent: 'rgba(0, 174, 236, 0.5)',

    colourForeground: '#FFFFFF',
    colourForegroundHover: '#F9F9F8',
    colourForegroundTransparent: 'rgba(255, 255, 255, 0.2)',

    colourInactive: '#7F7F7F',
    colourBackground: '#E5E5E5',
    colourDivider: '#D7D2CB',
    colourFaintBorder: '#8C9296',

    radius: '0.3rem',
    splitHeight: '50vh'
};
