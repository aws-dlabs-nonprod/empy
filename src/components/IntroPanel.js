import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';
import * as ActionTypes from '@constants/ActionTypes';
import Media from '@style/media';

import Button, { SecondaryButton, SecondaryButtonSmall } from '@components/Button';
import Loader from '@components/Loader';
import { SoulMachinesContext } from '@contexts/SoulMachines';

class InfoPanel extends Component {

    static contextType = SoulMachinesContext;

    constructor(props, context) {
        super(props, context);

        this.state = {
            isVideoStarted: false,
            isAudioStarted: false
        };

        this.handleVideoStart = this.handleVideoStart.bind(this);
        this.handleAudioStart = this.handleAudioStart.bind(this);
    }

    handleVideoStart() {
        this.props.startLoading();

        this.setState({ isVideoStarted: true });
        
        this.context.createScene(false)
            .then(() => {
                this.props.startVideo();
            });
    }

    handleAudioStart() {
        this.props.startLoading();

        this.setState({ isAudioStarted: true });

        this.context.createScene(true)
            .then(() => {
                this.props.startAudio();
            });
    }

    render() {
        const { hasCamera, hasMicrophone, isLoading, isBrowserSupported, allowAudioOnly } = this.props;
        const { isVideoStarted, isAudioStarted } = this.state;

        if (!isBrowserSupported) {
            return this.renderUnsupportedMessage();
        }

        return (
            <StyledInfoPanel>
                <div className="content">
                    <h1>Hi! I'm Wendy</h1>
                    
                    <p>If you turn on your camera and microphone I can see and hear you.</p>
                    <p>I find it easier to hear and focus on you if you are in a quiet space without too much background noise.</p>
                    <p>Ask me a question and I’ll do my best to answer it. I’m still learning so may not be able to answer all your questions.</p>
                    <p>Please don't tell me your personal details — I can’t give you any information relating to your personal banking or give you financial advice.</p>

                    <div className="actions">
                        { hasCamera && hasMicrophone &&
                            <Fragment>
                                <Button disabled={ isLoading } onClick={ this.handleVideoStart }>
                                    { isLoading && isVideoStarted ? <Loader /> : 'Start a video chat' }
                                </Button>
                                { allowAudioOnly &&
                                <SecondaryButton disabled={ isLoading } onClick={ this.handleAudioStart }>
                                    { isLoading && isAudioStarted ? <Loader /> : 'Start a voice chat' }
                                </SecondaryButton>
                                }
                            </Fragment>
                        }

                        { !hasCamera && hasMicrophone && allowAudioOnly &&
                            <Button disabled={ isLoading } onClick={ this.handleAudioStart }>
                                { isLoading ? <Loader /> : 'Start a voice chat' }
                            </Button>
                        }

                        {/*
                        Safari doesn't support webrtc video streams if you've not given webcam permissions
                        https://stackoverflow.com/a/53914556
                        */}
                        { ((!hasCamera && !hasMicrophone) || (!hasCamera && !allowAudioOnly)) &&
                            <p>It looks like you don't have a camera or microphone connected. Please connect one and try again.</p>
                        }

                    </div>
                </div>

                <SecondaryButtonSmall href="https://www.westpac.com.au/privacy/full-privacy-policy/" target="_blank">View privacy statement</SecondaryButtonSmall>
            </StyledInfoPanel>
        );
    }

    renderUnsupportedMessage() {
        return (
            <StyledInfoPanel>
                <div className="content">
                    <h1>Sorry</h1>
                    
                    <p>It looks like the browser or device you're using is not supported.</p>
                    <p>Wendy requires the latest version of Chrome, Firefox, Edge, or Safari on a desktop computer.</p>
                    <p>For mobile devices please use Apple iOS 11 or greater with Safari, or Android 4.4 or greater with Chrome or Firefox.</p>
                </div>
            </StyledInfoPanel>
        );
    }
}

const StyledInfoPanel = styled.div`
    align-items: flex-start;
    box-sizing: border-box;
    display: flex;
    padding: 7rem 0 0 0;
    position: absolute;
    top: 0;
    bottom: 0;
    width: 100%;
    z-index: 3;
    overflow: auto;

    ${Media.tablet`
        font-size: 1.4rem;
        padding: 11rem 0 4rem 0;
    `}

    ${Media.intro`
        align-items: center;
        padding: 0;
    `}

    .actions {
        background: ${props => props.theme.colourBackground};
        border-top: 0.1rem solid ${props => props.theme.colourDivider};
        bottom: 0;
        left: 0;
        padding: 1.5rem 1.5rem 6rem 1.5rem;
        position: fixed;
        right: 0;
        z-index: 2;

        >:last-child {
            margin-bottom: 0;
        }

        ${Media.tablet`
            background: none;
            position: relative;
            padding: 3rem 0 0 0;
            margin: 0;

            ${SecondaryButtonSmall} {
                display: none;
            }
        `}
    }

    .content {
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        margin: 0 1.5rem;
        padding: 0 0 13rem 0;

        ${Media.tablet`
            flex-basis: calc(41.6666666667% - 4.5rem);
            margin-left: calc(100% / 12 + 3rem);
            margin-right: 0;
            padding: 0;
        `}

        ${Media.desktop`
            flex-basis: calc(33.3333333333% - 4rem);
        `}

        ${Media.large`
            flex-basis: calc(25% - 4rem);
        `}
    }

    p:last-of-type {
        margin-bottom: 3rem;
    }

    ${SecondaryButtonSmall} {
        position: fixed;
        left: 1.5rem;
        bottom: 1.5rem;
        max-width: 20rem;
        margin: 0;

        ${Media.desktop`
            position: absolute;
        `}
    }
`;

function mapStateToProps(state) {
    return {
        hasCamera: state.hasCamera,
        hasMicrophone: state.hasMicrophone,
        isLoading: state.isLoading,
        isBrowserSupported: state.isBrowserSupported,
        allowAudioOnly: state.allowAudioOnly
    };
}

function mapDispatchToProps(dispatch) {
    return {
        startLoading: () => dispatch({
            type: ActionTypes.START_LOADING
        }),

        startVideo: () => dispatch({
            type: ActionTypes.SCENE_START_VIDEO
        }),

        startAudio: () => dispatch({
            type: ActionTypes.SCENE_START_AUDIO
        })
    };
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(InfoPanel);

