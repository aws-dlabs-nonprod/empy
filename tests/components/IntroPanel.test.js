import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import { Provider } from 'react-redux';
import 'jest-styled-components';
import theme from '@style/theme';
import { ThemeProvider } from 'styled-components';

import IntroPanel from '@components/IntroPanel';
import Button, { SecondaryButton } from '@components/Button';

const mockStore = configureStore([]);

describe('IntroPanel', () => {

    test('renders correctly', () => {
        const store = mockStore({});

        const component = shallow(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <IntroPanel />
                </ThemeProvider>
            </Provider>
        );

        expect(component).toMatchSnapshot();
    });

    test('handles no devices', () => {
        const store = mockStore({
            hasCamera: false,
            hasMicrophone: false,
            isBrowserSupported: true
        });

        const component = mount(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <IntroPanel />
                </ThemeProvider>
            </Provider>
        );

        expect(component.find(Button).length).toEqual(0);
        expect(component.text()).toContain('you don\'t have a camera or microphone');
    });

    test('handles only microphone', () => {
        const store = mockStore({
            hasCamera: false,
            hasMicrophone: true,
            isBrowserSupported: true
        });

        const component = mount(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <IntroPanel />
                </ThemeProvider>
            </Provider>
        );

        expect(component.find(Button).length).toEqual(1);
        expect(component.text()).toContain('Start a voice chat');
    });

    test('handles camera and microphone', () => {
        const store = mockStore({
            hasCamera: true,
            hasMicrophone: true,
            isBrowserSupported: true
        });

        const component = mount(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <IntroPanel />
                </ThemeProvider>
            </Provider>
        );

        expect(component.find(Button).length).toEqual(1);
        expect(component.find(SecondaryButton).length).toEqual(1);
        expect(component.text()).toContain('Start a video chat');
        expect(component.text()).toContain('Start a voice chat');
    });

    test('handles unsupported browser', () => {
        const store = mockStore({
            hasCamera: true,
            hasMicrophone: true,
            isBrowserSupported: false
        });

        const component = mount(
            <Provider store={ store }>
                <ThemeProvider theme={ theme }>
                    <IntroPanel />
                </ThemeProvider>
            </Provider>
        );

        expect(component.text()).toContain('It looks like the browser or device you\'re using is not supported.');
    });
});
