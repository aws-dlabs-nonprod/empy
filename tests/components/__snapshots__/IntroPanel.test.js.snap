// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntroPanel renders correctly 1`] = `
<ThemeProvider
  theme={
    Object {
      "colourBackground": "#E5E5E5",
      "colourDivider": "#D7D2CB",
      "colourFaintBorder": "#8C9296",
      "colourForeground": "#FFFFFF",
      "colourForegroundHover": "#F9F9F8",
      "colourForegroundTransparent": "rgba(255, 255, 255, 0.2)",
      "colourInactive": "#7F7F7F",
      "colourPrimary": "#D5002B",
      "colourPrimaryHover": "#e24d6b",
      "colourPrimaryTransparent": "rgba(0, 174, 236, 0.5)",
      "radius": "0.3rem",
      "splitHeight": "50vh",
      "textBody": "#1A1A1A",
      "textDark": "#621A4B",
      "textPrimary": "#2D373E",
      "textSecondary": "#E7E6E6",
    }
  }
>
  <Connect(InfoPanel) />
</ThemeProvider>
`;
